File: backend/app/api/v1/kubernetes.py
from fastapi import APIRouter, Depends, Query

from app.api.deps import get_k8s_client
from app.models.kubernetes import PodStatus, DeploymentStatus, K8sEvent
from app.security import validate_identifier
from app.services.kubernetes_client import KubernetesClient

router = APIRouter(prefix="/kubernetes", tags=["kubernetes"])


@router.get("/pods", response_model=list[PodStatus])
async def get_pods(
    namespace: str | None = Query(None, max_length=128),
    k8s: KubernetesClient = Depends(get_k8s_client),
):
    if namespace:
        namespace = validate_identifier(namespace, "namespace")
    return await k8s.list_pods(namespace=namespace)


@router.get("/deployments", response_model=list[DeploymentStatus])
async def get_deployments(
    namespace: str | None = Query(None, max_length=128),
    k8s: KubernetesClient = Depends(get_k8s_client),
):
    if namespace:
        namespace = validate_identifier(namespace, "namespace")
    return await k8s.list_deployments(namespace=namespace)


@router.get("/events", response_model=list[K8sEvent])
async def get_events(
    namespace: str | None = Query(None, max_length=128),
    k8s: KubernetesClient = Depends(get_k8s_client),
):
    if namespace:
        namespace = validate_identifier(namespace, "namespace")
    return await k8s.get_events(namespace=namespace)

File: frontend/src/api/kubernetes.ts
import { api } from './client';
import type { PodStatus, DeploymentStatus, K8sEvent } from '../types';

export async function fetchPods(params?: { namespace?: string }): Promise<PodStatus[]> {
  const { data } = await api.get<PodStatus[]>('/api/v1/kubernetes/pods', { params });
  return data;
}

export async function fetchDeployments(params?: { namespace?: string }): Promise<DeploymentStatus[]> {
  const { data } = await api.get<DeploymentStatus[]>('/api/v1/kubernetes/deployments', { params });
  return data;
}

export async function fetchK8sEvents(params?: { namespace?: string }): Promise<K8sEvent[]> {
  const { data } = await api.get<K8sEvent[]>('/api/v1/kubernetes/events', { params });
  return data;
}
